jetty-websocket/websocket-common/src/main/java/org/eclipse/jetty/websocket/common/io/AbstractWebSocketConnection.java

```diff
@@ -432,112 +234,109 @@ public boolean isReading()
-                }
-                if (LOG.isDebugEnabled())
-                {
-                    LOG.debug("OPEN: normal fillInterested");
-                }
-                // TODO: investigate what happens if a failure occurs during prefill, and an attempt to write close fails,
-                // should a fill interested occur? or just a quick disconnect?
-                fillInterested();
-                break;
-            case CLOSED:
-                if (LOG_CLOSE.isDebugEnabled())
-                    LOG_CLOSE.debug("CLOSED - wasAbnormalClose: {}", ioState.wasAbnormalClose());
-                if (ioState.wasAbnormalClose())
-                {
-                    // Fire out a close frame, indicating abnormal shutdown, then disconnect
-                    CloseInfo abnormal = new CloseInfo(StatusCode.SHUTDOWN,"Abnormal Close - " + ioState.getCloseInfo().getReason());
-                    outgoingFrame(abnormal.asFrame(),new OnDisconnectCallback(false),BatchMode.OFF);
-                }
-                else
-                {
-                    // Just disconnect
-                    this.disconnect(false);
-                }
-                break;
-            case CLOSING:
-                if (LOG_CLOSE.isDebugEnabled())
-                    LOG_CLOSE.debug("CLOSING - wasRemoteCloseInitiated: {}", ioState.wasRemoteCloseInitiated());
-                // First occurrence of .onCloseLocal or .onCloseRemote use
-                if (ioState.wasRemoteCloseInitiated())
+            @Override
+            public void succeed()
+            {
+                parser.release(frame);
+                if(!result.compareAndSet(false,true))
                 {
-                    CloseInfo close = ioState.getCloseInfo();
-                    // reply to close handshake from remote
-                    outgoingFrame(close.asFrame(),new OnCloseLocalCallback(new OnDisconnectCallback(true),close),BatchMode.OFF);
+                    // callback has been notified asynchronously
+                    fillAndParse();
                 }
-            default:
-                break;
+            }
+            
+            @Override
+            public void fail(Throwable cause)
+            {
+                parser.release(frame);
+                
+                // notify session & endpoint
+                notifyError(cause);
+            }
+        });
+        
+        if(result.compareAndSet(false, true))
+        {
+            // callback hasn't been notified yet
+            return false;
         }
+        
+        return true;
     }
-
+    
     @Override
     public void onFillable()
     {
-        if (LOG.isDebugEnabled())
-            LOG.debug("{} onFillable()",policy.getBehavior());
-        stats.countOnFillableEvents.incrementAndGet();
-
-        ByteBuffer buffer = bufferPool.acquire(getInputBufferSize(),true);
-
+        networkBuffer = bufferPool.acquire(getInputBufferSize(),true);
```

Comment:
not sure I follow, but it'll probably become clear once the change to prefill the networkbuffer from the upgrade prefill is made.